src/table_ops.jl

```diff
@@ -0,0 +1,307 @@
+- `dims`: One or more dimensions to be inferred. If no dimensions are specified, then `guess_dims` will default
+to any available dimensions in the set `(:X, :Y, :Z, :Ti, :Band)`. Dimensions can be given as either a singular
+value or as a `Pair` with both the dimensions and corresponding order. The order will be inferred from the data
+when none is given. This should work for sorted coordinates, but will not be sufficient when the table's rows are
+out of order.
+  
+# Keyword Arguments
+- `precision`: Specifies the number of digits to use for guessing dimensions (default = `6`).
+
+# Returns
+A tuple containing the inferred dimensions from the table.
+
+# Example
+```julia
+julia> xdims = X(LinRange{Float64}(610000.0, 661180.0, 2560));
+
+julia> ydims = Y(LinRange{Float64}(6.84142e6, 6.79024e6, 2560));
+
+julia> bdims = Dim{:Band}([:B02, :B03, :B04]);
+
+julia> d = DimArray(rand(UInt16, 2560, 2560, 3), (xdims, ydims, bdims));
+
+julia> t = DataFrame(d);
+
+julia> t_rand = Random.shuffle(t);
+
+julia> dims(d)
+↓ X    Sampled{Float64} LinRange{Float64}(610000.0, 661180.0, 2560) ForwardOrdered Regular Points,
+→ Y    Sampled{Float64} LinRange{Float64}(6.84142e6, 6.79024e6, 2560) ReverseOrdered Regular Points,
+↗ Band Categorical{Symbol} [:B02, :B03, :B04] ForwardOrdered
+
+julia> DD.guess_dims(t)
+↓ X    Sampled{Float64} LinRange{Float64}(610000.0, 661180.0, 2560) ForwardOrdered Regular Points,
+→ Y    Sampled{Float64} LinRange{Float64}(6.84142e6, 6.79024e6, 2560) ReverseOrdered Regular Points,
+↗ Band Categorical{Symbol} [:B02, :B03, :B04] ForwardOrdered
+
+julia> DD.guess_dims(t, (X, Y, :Band))
+↓ X    Sampled{Float64} LinRange{Float64}(610000.0, 661180.0, 2560) ForwardOrdered Regular Points,
+→ Y    Sampled{Float64} LinRange{Float64}(6.84142e6, 6.79024e6, 2560) ReverseOrdered Regular Points,
+↗ Band Categorical{Symbol} [:B02, :B03, :B04] ForwardOrdered
+
+julia> DD.guess_dims(t_rand, (X => DD.ForwardOrdered(), Y => DD.ReverseOrdered(), :Band => DD.ForwardOrdered()))
+↓ X    Sampled{Float64} LinRange{Float64}(610000.0, 661180.0, 2560) ForwardOrdered Regular Points,
+→ Y    Sampled{Float64} LinRange{Float64}(6.84142e6, 6.79024e6, 2560) ReverseOrdered Regular Points,
+↗ Band Categorical{Symbol} [:B02, :B03, :B04] ForwardOrdered
+```
+"""
+guess_dims(table; kw...) = guess_dims(table, filter(x -> x in Tables.columnnames(table), (:X,:Y,:Z,:Ti,:Band)); kw...)
+guess_dims(table, dims::Tuple; kw...) = map(dim -> guess_dims(table, dim; kw...), dims)
+guess_dims(table, dim; precision=6) = _guess_dims(_get_column(table, dim), dim, precision)
+
+_guess_dims(coords::AbstractVector, dim::DD.Dimension, args...) = dim
+_guess_dims(coords::AbstractVector, dim::Type{<:DD.Dimension}, args...) = _guess_dims(coords, DD.name(dim), args...)
+_guess_dims(coords::AbstractVector, dim::Pair, args...) = _guess_dims(coords, first(dim), last(dim), args...)
+function _guess_dims(coords::AbstractVector, dim::Symbol, precision::Int)
```

Comment:
Should this be `AbstractVector{Union{Number,DateTime}` ?

I'm wondering what happens to strings, symbols and other objects that need to go in `Categorical` lookups.

Reply:
It should work with almost any type. If the coordinates are non-numerical, then we will internally dispatch on the following methods:

```julia
# Extract all unique coordinates from the given vector
_unique_vals(coords::AbstractVector, precision::Int) = _round_dim_val.(coords, precision) |> unique

# Round dimension value within the specified precision
_round_dim_val(x, ::Int) = x

# Determine if the given coordinates are forward ordered, reverse ordered, or unordered
function _guess_dim_order(coords::AbstractVector)
    if issorted(coords)
        return DD.ForwardOrdered()
    elseif issorted(coords, rev=true)
        return DD.ReverseOrdered()
    else
        return DD.Unordered()
    end
end

# Estimate the span between consecutive coordinates
_guess_dim_span(::AbstractVector, ::DD.Order, ::Int) = DD.Irregular()
```

`_unique_vals` will just return all unique values, where `_round_dim_val` is just the identity function for non-numerical coordinates. 

`_guess_dim_order` should work for anything that can be sorted, which is the case for both `String` and `Symbol`.

`_guess_dim_span` will return `DD.Irregular()` for non-numerical coordinates.

Reply:
We used to need a try catch for `issorted` in case `<` is not defined for a type

Reply:
This should let us handle data that can't be sorted:

```julia
function _guess_dim_order(coords::AbstractVector)
    try
        if issorted(coords)
            return DD.ForwardOrdered()
        elseif issorted(coords, rev=true)
            return DD.ReverseOrdered()
        else
            return DD.Unordered()
        end
    catch 
        return DD.Unordered()
    end
end
```

And this should retrieve the span from `Date` and `DateTime` objects:

```julia
function _guess_dim_span(coords::AbstractVector{<:Dates.AbstractTime}, ::DD.Ordered, precision::Int)
    steps = (@view coords[2:end]) .- (@view coords[1:end-1])
    span = argmin(abs, steps)
    return all(isinteger, round.(steps ./ span, digits=precision)) ? DD.Regular(span) : DD.Irregular()
end
```

However, there seems to be a problem with constructing a `LinRange` from `Date` objects:

```julia
julia> vals = [Date("2022-11-16") + Day(i * 7) for i in 0:4];

julia> LinRange(first(vals), last(vals), 5)
5-element LinRange{Day, Int64}:
Error showing value of type LinRange{Day, Int64}:
ERROR: InexactError: Int64(553856.25)
```
Thus, I'm not sure how we should construct a `Dimension` with regularly spaced dates.

Reply:
A StepRange works for Dates. Probably we should use StepRangeLen instead of LinRange where possible anyway

Reply:
That works. Do you want to use `StepRangeLen` for numerical coordinates or just dates?

Here's the result:
```julia
julia> xdims = X(LinRange{Float64}(610000.0, 661180.0, 2560));

julia> ydims = Y(LinRange{Float64}(6.84142e6, 6.79024e6, 2560));

julia> bdims = Dim{:Band}([:B02, :B03, :B04]);

julia> tdims = Dim{:Ti}([d1 + Day(i * 7) for i in 0:4]);

julia> d = DimArray(rand(UInt16, 2560, 2560, 3, 5), (xdims, ydims, bdims, tdims));

julia> t = DataFrame(d);

julia> DD.guess_dims(t)
↓ X    Sampled{Float64} 610000.0:20.0:661180.0 ForwardOrdered Regular Points,
→ Y    Sampled{Float64} 6.84142e6:-20.0:6.79024e6 ReverseOrdered Regular Points,
↗ Ti   Sampled{Date} Date("2024-11-18"):Day(7):Date("2024-12-16") ForwardOrdered Regular Points,
⬔ Band Categorical{Symbol} [:B02, :B03, :B04] ForwardOrdered
```

Reply:
It's probably better for everything, except a bit slower. It didn't exist when I first wrote this package and uses of LinRange are just legacy from that.